### DISCUSSION TITLE
Feature File Execution in some specific order

### DISCUSSION THREAD

**Discussion by adityakumar1990** [behave/behave#999] [Ideas]
Hi Guys,
I have one scenario to implement 
suppose there are 10 features and each features can have more than one scenario .
After complete execution of given step of a scenario , the behave framework should pick the given for other scenrio in the same feature and if no other scenario is there then it should pick another feature given steps .Once all the given steps for all features is completed then it should wait for one condition(say db entry) to be true then it start executing then step for each feature file .

i.e calling one feature grom another running feature and keep calling all the features , once done then wait for a sometime and then come back to first feature and start executing all features again .

is there any way to implement this in behave ?

**Comment by jenisys** [comment]
Mmh, I think I understand partly what you are trying to achieve.
BUT: You should provide a motivating example to avoid confusion in the further discussion (with 2 feature files) in Gherkin, each having at least one or two scenarios). In addition, you should describe how the execution of steps should occur in your opinion.

NOTES:
* `behave` test runner will not wait for a condition (reason: it is generic).
  BUT: You can wait on something in your step/hook/fixture implementation(s).
* IF I UNDERSTAND YOU CORRECTLY: Executing only the given steps of scenarios before the other steps are executed is a bad idea. It indicates that your scenarios are correlated which should not be the case. In addition, it hints that your scenario setup (with given step) is partly incomplete for some scenarios.

**Comment by adityakumar1990** [comment]
```gherkin
# — FILE: features/feature_001.feature
Feature: f1
 Scenario: s1
  Given g1
  When  w1
  Then  t1
```

```gherkin
# — FILE: features/feature_002.feature
Feature: f2
 Scenario: s2
  Given g2
  When  w2
  Then  t2
```

No I want my execution to follow:

```
run s1(g1&w1)
now run s2(g2&w2)
then run s1(t1) and s2(t2)
```

There is no dependency between any features and steps:
* you can think g1 as read data from datatable ,w1 to insert data in tabl1 
* g2 as read data from some datatable ,w2 to insert data in tabl2 , 
  Then one dev script is there which is running in every 30 mins , 
  Once the script execution completes(we donot have any control on script execution and we cannot call it explicitly also) 
* then run t1 and t2 .

I hope now my requirement is clear.

**Reply by adityakumar1990** [reply]
feature_001.txt
Feature: f1
 Scenario: s1
  Given g1
  When  w1
  Then  t1

feature_002.txt
Feature: f2
 Scenario: s2
  Given g2
  When  w2
  Then  t2


now I want my execution to follow -->
   run s1(g1&w1) now run s2(g2&w2) then run s1(t1) and s2(t2)

there is no dependency between any features and steps . you can think g1 as read data from datatable ,w1 to insert data in tabl1 
g2 as read data from some datatable ,w2 to insert data in tabl2 , then one dev script is there which is running in every 30 mins , once the script execution completes(we donot have any control on script execution and we cannot call it explicitly also) then run t1 and t2 .

hope now my requirement is clear

**Comment by jenisys** [comment]
Mmh, the execution of each scenario is independent from the each other scenario (in: `behave`).
What you can do in the implementation of your steps is the following:

* At the beginning of step `t1` and/or `t2` you check that a condition is met (or that you receive an event when this condition becomes true). Therefore, you basically wait inside of these steps that a certain condition is met (before the core functionality of this step is executed)

OTHERWISE:
* Split-up into 2 steps for given/when seems unnecessary for your use case.
  PROBABLY BETTER: Combine the functionality of both steps in one `given step`.
* Avoid during automation of your tests to wait 30 minutes (if that is the case)
* You dev-script needs somehow to fire an event that indicates to `then steps` that the condition to continue running is met
* AVOID DEADLOCKS: While waiting in the `then steps` that the event is received or condition is met (which may never occur). Use TIMEOUTs to recover from that (which causes the `then-step` to fail.
